PolyEngine/Core/Src/pe/core/storage/Queue.hpp

```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+
+			bool operator==(const ConstIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ConstIterator& rhs) const { return idx != rhs.idx; }
+
+			const T& operator*() const { return ref.m_data.at(idx); }
+
+			ConstIterator& operator++() { idx = ref.getNextIdx(idx); return *this; }
+			ConstIterator operator++(int) { ConstIterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+			ConstIterator& operator--() { idx = ref.getNextIdx(idx); return *this; }
+			ConstIterator operator--(int) { ConstIterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+		private:
+			ConstIterator(const Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			const Queue<T>& ref;
+			size_t idx;
+
+		};
+
+		class ReverseIterator : public BaseObject<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const ReverseIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ReverseIterator& rhs) const { return idx != rhs.idx; }
+
+			T& operator*() const { return ref.m_data.at(idx); }
+
+			ReverseIterator& operator++() { increment(); return *this; }
+			ReverseIterator operator++(int) { ReverseIterator ret(ref, idx); increment(); return ret; }
+			ReverseIterator& operator--() { idx = ref.getPrevIdx(idx); return *this; }
+			ReverseIterator operator--(int) { ReverseIterator ret(ref, idx); idx = ref.getPrevIdx(idx); return ret; }
+		private:
+			ReverseIterator(Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			void increment()
+			{
+				if(idx == ref.m_head)
+					idx = ref.m_tail;
+				else 
+					idx = ref.getPrevIdx(idx);
+			}
+			Queue<T>& ref;
+			size_t idx;
+		};
+
+		class ConstReverseIterator : public BaseObject<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const ConstReverseIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ConstReverseIterator& rhs) const { return idx != rhs.idx; }
+
+			const T& operator*() const { return ref.m_data.at(idx); }
+
+			ConstReverseIterator& operator++() { increment(); return *this; }
+			ConstReverseIterator operator++(int) { ConstReverseIterator ret(ref, idx); increment(); return ret; }
+			ConstReverseIterator& operator--() { idx = ref.getPrevIdx(idx); return *this; }
+			ConstReverseIterator operator--(int) { ConstReverseIterator ret(ref, idx); idx = ref.getPrevIdx(idx); return ret; }
+		private:
+			ConstReverseIterator(const Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			void increment()
+			{
+				if(idx == ref.m_head)
+					idx = ref.m_tail;
+				else 
+					idx = ref.getPrevIdx(idx);
+			}
```

Comment:
The way to implement interators in a circular buffer is too always keep two elements in the queue:
- element before head - which is == queue.rend()
- element after end - which is == queue.end()
I'm not sure whether the iterators work correctly in the current setup. For sure we need good test coverage here.

Reply:
Currently there is a workaround implemented, please review if it is suitable. It is hidden from outside world, and test do cover it.

Reply:
proper solution implemented in newest commits